core/trino-main/src/test/java/io/trino/operator/BenchmarkGroupByHashOnTpch.java

```diff
@@ -0,0 +1,480 @@
+                results.add(work.getResult()); // Pretend the results are used
+            }
+            while (!finished);
+        }
+    }
+
+    public interface BlockWriter
+    {
+        void write(BlockBuilder blockBuilder, int positionCount, int cardinality, long randomSeed);
+    }
+
+    public enum ColumnType
+    {
+        BIGINT(BigintType.BIGINT, (blockBuilder, positionCount, cardinality, seed) -> {
+            Random r = new Random(seed);
+            for (int i = 0; i < positionCount; i++) {
+                blockBuilder.writeLong((r.nextLong() >>> 1) % cardinality); // Only positives
+            }
+        }),
+        INT(IntegerType.INTEGER, (blockBuilder, positionCount, cardinality, seed) -> {
+            Random r = new Random(seed);
+            for (int i = 0; i < positionCount; i++) {
+                blockBuilder.writeInt(r.nextInt(cardinality));
+            }
+        }),
+        DOUBLE(DoubleType.DOUBLE, (blockBuilder, positionCount, cardinality, seed) -> {
+            Random r = new Random(seed);
+            for (int i = 0; i < positionCount; i++) {
+                blockBuilder.writeLong((r.nextLong() >>> 1) % cardinality); // Only positives
+            }
+        }),
+        VARCHAR_10(VarcharType.VARCHAR, (blockBuilder, positionCount, cardinality, seed) -> {
+            writeVarchar(blockBuilder, positionCount, cardinality, seed, 10);
+        }),
+        VARCHAR_15(VarcharType.VARCHAR, (blockBuilder, positionCount, cardinality, seed) -> {
+            writeVarchar(blockBuilder, positionCount, cardinality, seed, 15);
+        }),
+        VARCHAR_25(VarcharType.VARCHAR, (blockBuilder, positionCount, cardinality, seed) -> {
+            writeVarchar(blockBuilder, positionCount, cardinality, seed, 25);
+        }),
+        VARCHAR_40(VarcharType.VARCHAR, (blockBuilder, positionCount, cardinality, seed) -> {
+            writeVarchar(blockBuilder, positionCount, cardinality, seed, 40);
+        }),
+        VARCHAR_117(VarcharType.VARCHAR, (blockBuilder, positionCount, cardinality, seed) -> {
+            writeVarchar(blockBuilder, positionCount, cardinality, seed, 117);
+        }),
+        CHAR_1(CharType.createCharType(1), (blockBuilder, positionCount, cardinality, seed) -> {
+            Random r = new Random(seed);
+            for (int i = 0; i < positionCount; i++) {
+                byte value = (byte) r.nextInt(cardinality);
+                while (value == ' ') {
+                    value = (byte) r.nextInt(cardinality);
+                }
+                CharType.createCharType(1).writeSlice(blockBuilder, Slices.wrappedBuffer(value));
+            }
+        }),
+        /**/;
+
+        private static void writeVarchar(BlockBuilder blockBuilder, int positionCount, int cardinality, long seed, int maxLength)
+        {
+            Random r = new Random(seed);
+            for (int i = 0; i < positionCount; i++) {
+                Random r2 = new Random(r.nextInt(cardinality));
```

Comment:
why two `Random`s?

Reply:
It's a way of generating only `n` different sets of data. Ideally I would just generate `n` values, put them in some array and then choose out of those values. But using a random with a fixed seed will produce the same values without the need to keep them anywhere. 
This is not performant but who cares.

Reply:
I don't get it, using a single Radom would not generate `n` different sets?

Reply:
I don't see how would you like to make it work. We'll talk offline about it.